Add documentation for dataset factories feature

ankatiyar opened the pull request:
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description
<!-- Why was this PR created? -->
Resolve #2666 
To be merged with/after https://github.com/kedro-org/kedro/pull/2635

## Development notes
<!-- What have you changed, and how has this been tested? -->
Documentation for the dataset factories feature to the "The Data Catalog" page. 


## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,7 +461,75 @@ airplanes:
 
 In this example, the default `csv` configuration is inserted into `airplanes` and then the `load_args` block is overridden. Normally, that would replace the whole dictionary. In order to extend `load_args`, the defaults for that block are then re-inserted.
 
+## Load multiple dataset with similar configuration using dataset factory patterns
```
```suggestion
## Load multiple datasets with similar configuration using dataset factories
```

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,7 +461,75 @@ airplanes:
 In this example, the default `csv` configuration is inserted into `airplanes` and then the `load_args` block is overridden. Normally, that would replace the whole dictionary. In order to extend `load_args`, the defaults for that block are then re-inserted.
 
+## Load multiple dataset with similar configuration using dataset factory patterns
+For catalog entries that share a lot of the same configurations, you could also use the dataset factory pattern introduced in Kedro 0.18.11. This syntax allows you to
```
```suggestion
For catalog entries that share a lot of the same configurations, you could also use the dataset factories introduced in Kedro 0.18.11. This syntax allows you to
```

merelcht commented on docs/source/data/data_catalog.md:
````diff
@@ -461,7 +461,75 @@ airplanes:
+  type: pandas.CSVDataSet
+  filepath: data/01_raw/{name}_data.csv
+```
+When `factory_data` or `process_data` is used in your pipeline, it would be matched to the factory pattern above. Two things to keep in mind are that the factory pattern name should be enclosed in
````
```suggestion
When `factory_data` or `process_data` is used in your pipeline, it will be matched to the factory pattern above. Two things to keep in mind are that the factory pattern name should be enclosed in
```

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,7 +461,75 @@ airplanes:
+When `factory_data` or `process_data` is used in your pipeline, it would be matched to the factory pattern above. Two things to keep in mind are that the factory pattern name should be enclosed in
+quotes and that the placeholders used in the catalog entry body should be present in the factory pattern name.
+
+You can also use multiple placeholders in the same pattern. For example, consider the following catalog where the dataset entries share `type`, `file_format` and `save_args`:
```
```suggestion
You can use multiple placeholders in the same pattern. For example, consider the following catalog where the dataset entries share `type`, `file_format` and `save_args`:
```

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,7 +461,75 @@ airplanes:
+that would be picked over any factory pattern, then the pattern with the highest number of matching characters, and finally the pattern with the highest number of placeholders in the pattern name.
+
+
+You can use dataset factory pattern to define a catch-all dataset factory pattern which would overwrite the default `MemoryDataSet` creation.
```
```suggestion
You can use dataset factories to define a catch-all pattern which will overwrite the default `MemoryDataSet` creation.
```

merelcht reviewed (commented):
This is a great start for the docs! 

Some comments: I think we should refer to the feature as "dataset factories" instead of "dataset factory patterns". IMO the complete dataset block, e.g.: 

```
"{namespace}.{dataset_name}#csv":
   type: pandas.CSVDataSet
   filepath: data/{namespace}/{dataset_name}.csv
```
is the dataset factory and the name: `"{namespace}.{dataset_name}#csv"` is the pattern.

Several users mentioned in the interview that they would like to see more examples, specifically how the patterns are referenced in the pipeline. The data catalog page is super long and I know @stichbury is keen to restructure it, so I can imagine dataset factories will perhaps get a separate page, but for now I think it's important to include lots of examples even if that increases this page more.

stichbury commented on docs/source/data/data_catalog.md:
````diff
@@ -461,7 +461,75 @@ airplanes:
+  filepath: data/01_raw/{name}_data.csv
+```
+When `factory_data` or `process_data` is used in your pipeline, it would be matched to the factory pattern above. Two things to keep in mind are that the factory pattern name should be enclosed in
+quotes and that the placeholders used in the catalog entry body should be present in the factory pattern name.
````
```suggestion
quotes and the placeholders used in the catalog entry body should be included in the factory pattern name.
```

stichbury reviewed (approved):
This looks good to me, with the suggestions. I made a few edits to @merelcht's to add a few more minor fixes but they're all good and this is such a great feature. I think we could do a mini blog post about this with some examples, maybe? WDYT?

merelcht commented:
>  I think we could do a mini blog post about this with some examples, maybe? WDYT?

A blog post would be great! I think we should also do a show and tell session for our users with this feature to drive adoption. There was so much interest for the interviews so I can imagine lots of people would like to know about this.

ankatiyar commented:
@merelcht & @stichbury Thanks for the feedback! I've incorporated all the edits. I've also added more examples and structured them under separate subheadings. Would you mind taking another look and letting me know if I've missed anything? :)

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,6 +461,154 @@ airplanes:
+
+### Example 2: Generalise datasets with similar types into one dataset factory
+You can also combine all the datasets with the same type and configuration details. For example, consider the following
+catalog with three datasets named `reviews`, `shuttles` and `reviews` of the type `pandas.CSVDataSet`:
```
```suggestion
catalog with three datasets named `reviews`, `shuttles` and `companies` of the type `pandas.CSVDataSet`:
```

merelcht commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,154 @@ airplanes:
+        ]
+    )
+```
+### Example 3: Generalise datasets with similar types using into one dataset factory with multiple placeholders
````
```suggestion
### Example 3: Generalise datasets of the same type using namespaces into one dataset factory with multiple placeholders
```

merelcht commented on docs/source/data/data_catalog.md:
```diff
@@ -461,6 +461,154 @@ airplanes:
+quotes to avoid YAML parsing errors.
+
+
+### Example 2: Generalise datasets with similar types into one dataset factory
```
```suggestion
### Example 2: Generalise datasets of the same type into one dataset factory
```

merelcht commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,154 @@ airplanes:
+```
+### Example 3: Generalise datasets with similar types using into one dataset factory with multiple placeholders
+
+You can use multiple placeholders in the same pattern. For example, consider the following catalog where the dataset entries share `type`, `file_format` and `save_args`:
````
```suggestion
You can use multiple placeholders in the same pattern. For example, consider the following catalog where the dataset entries share `type`, `file_format` and `save_args` and all datasets are part of a namespace:
```

merelcht commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,154 @@ airplanes:
+```yaml
+"{default_dataset}":
+  type: pandas.CSVDataSet
+  filepath: <directory>/{default_dataset}.csv
````
```suggestion
  filepath: data/{default_dataset}.csv
```
I think having `<directory>` in there might confuse people that that's also special syntax for the dataset factories.

merelcht reviewed (approved):
This looks great now! 👍

astrojuanlu commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,204 @@ airplanes:
+```
+The datasets in this catalog can be generalised to the following dataset factory:
+```yaml
+"{name}_data":
````
Since this is the first example, I'm wondering if it would be good to add an "info" admonition saying

"If you don't use a suffix the default dataset will be overridden, see [this section below]"

Wdyt?

ankatiyar replied:
I kept the first example basic enough to get started but the second example actually shows a situation where you might need to add a suffix/prefix. I added a small explanation there as to why.

astrojuanlu reviewed (approved):
Looks great @ankatiyar! Just added one nitpick, but approved anyway

noklam commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,205 @@ airplanes:
+You can also combine all the datasets with the same type and configuration details. For example, consider the following
+catalog with three datasets named `reviews`, `shuttles` and `companies` of the type `pandas.CSVDataSet`:
+```yaml
+shuttles:
+  type: pandas.CSVDataSet
+  filepath: data/01_raw/shuttles.csv
````
I guess this is just an example, but it reuses a lot of `spaceflights` tutorial, if I create a spaceflights project and do this I get error because I only have `shuttles.xlsx` but not `shuttles.csv`

ankatiyar replied:
Yeah, I changed it to be csv just for the example but you are right, it might cause confusion. Changed it!

astrojuanlu commented on docs/source/data/data_catalog.md:
````diff
@@ -461,6 +461,205 @@ airplanes:
+```
+These datasets can be combined into the following dataset factory:
+```yaml
+"{dataset_name}#csv":
````
Yeah maybe something else like `@`, `$`, `%`? (Sorry if this has been treated already elsewhere)

ankatiyar replied:
It's not something we recommend afaik, I've tried to use different delimiters for each example, underscore, hash symbol, period etc
